src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -354,7 +354,7 @@ private GeneratorDriver BuildFinalCompilation(Compilation compilation, out Compi
         private static GeneratorState SetGeneratorException(CommonMessageProvider provider, GeneratorState generatorState, ISourceGenerator generator, Exception e, DiagnosticBag? diagnosticBag, bool isInit = false)
         {
             var message = isInit ? provider.WRN_GeneratorFailedDuringInitialization : provider.WRN_GeneratorFailedDuringGeneration;
-            var diagnostic = Diagnostic.Create(provider, message, generator.GetType().Name);
+            var diagnostic = Diagnostic.Create(provider, message, generator.GetType().Name, e.GetType().Name, e.Message, e.ToString());
```

Comment:
>e.ToString() [](start = 121, length = 12)

Please pass `e` rather than `e.ToString()`, to avoid computing the full string until the message is consumed. And don't we need to delay to allow the consumer to use a custom `IFormatProvider`?

Reply:
So, I originally planned to do this, but when you try and create a Diagnostic with the exception itself, you hit a path that checks the argument is serializable, and fails if it isn't: [https://github.com/dotnet/roslyn/blob/5f5bf37e10c03250ff627dbe70e9d15657d5d9e6/src/Compilers/Core/Portable/Diagnostic/DiagnosticInfo.cs#L60](https://github.com/dotnet/roslyn/blob/5f5bf37e10c03250ff627dbe70e9d15657d5d9e6/src/Compilers/Core/Portable/Diagnostic/DiagnosticInfo.cs#L60)

For analyzers we directly `.ToString()`s the exception for the description

---
In reply to: [471831098](https://github.com/dotnet/roslyn/pull/46804#discussion_r471831098) [](ancestors = 471831098)

Reply:
It seems surprising we'd call `Exception.ToString()` either here or in the analyzer code path, particularly since the text is for the diagnostic description and it's unclear (to me at least) how or when the description is used. Should we simply allow`Exception` instances as arguments in `DiagnosticInfo`?

Given that analyzers already call `.ToString()`, perhaps we should log a bug to consider this separately.

---
In reply to: [471860114](https://github.com/dotnet/roslyn/pull/46804#discussion_r471860114) [](ancestors = 471860114,471831098)